instance/models/openedx_instance.py

```diff
@@ -52,73 +62,105 @@ class OpenEdXInstance(Instance, OpenEdXAppConfiguration, OpenEdXDatabaseMixin, O
     """
 
     # Most settings/fields are inherited from OpenEdXAppConfiguration
-    sub_domain = models.CharField(max_length=50)
-    base_domain = models.CharField(max_length=50, blank=True)
+
+    # Internal domains are controlled by us and their DNS records are automatically set to point to the current active
+    # appserver. They are generated from a unique prefix (given as 'sub_domain' in instance factories) and the value of
+    # DEFAULT_INSTANCE_BASE_DOMAIN at instance creation time. They cannot be blank and are normally never changed after
+    # the instance is created.
+    # External domains on the other hand are controlled by the customer and are optional. We use external domains in
+    # preference to internal domains when displaying links to the instance in the UI and when passing domain related
+    # settings to ansible vars when provisioning appservers.
+    # The `domain`, `lms_preview_domain`, and `studio_domain` methods are useful when you need to access the external
+    # domain if present, otherwise fall back to the corresponding internal domain.
+    internal_lms_domain = models.CharField(max_length=100, blank=False, unique=True)
+    internal_lms_preview_domain = models.CharField(max_length=100, blank=False, unique=True)
+    internal_studio_domain = models.CharField(max_length=100, blank=False, unique=True)
+
+    external_lms_domain = models.CharField(max_length=100, blank=True)
+    external_lms_preview_domain = models.CharField(max_length=100, blank=True)
+    external_studio_domain = models.CharField(max_length=100, blank=True)
 
     active_appserver = models.OneToOneField(
         OpenEdXAppServer, null=True, blank=True, on_delete=models.SET_NULL, related_name='+',
     )
 
     class Meta:
-        unique_together = ('base_domain', 'sub_domain')
         verbose_name = 'Open edX Instance'
 
+    def __init__(self, *args, **kwargs):
+        """
+        OpenEdXInstance constructor.
+
+        The constructor is overridden to optionally accept a 'sub_domain' parameter instead of a full
+        value for 'internal_lms_domain'. When 'sub_domain' is provided, the 'internal_lms_domain' field is automatically
+        generated from from the value of 'sub_domain' and the DEFAULT_INSTANCE_BASE_DOMAIN setting.
+        """
+        sub_domain = None
```

Comment:
@itsjeyd Ah yes, that's a leftover from previous iteration before I refactored the code. Thanks for spotting it.